windowing.py

```diff
@@ -205,3 +207,68 @@ def cropped_np(np_sample, start, end):
     end_index = np.where(nanos == end)
     cropped = [np_sample[:,  start_index[0][0]:end_index[0][0] + 1]]
     return cropped
+
+
+def get_start_time(lift_windows, rows, window_size, stride, variance_threshold):
+    """
+        Description: Returns the precise start time if it is not None or returns the start found 
+                     from find_initial_lift_times 
+
+        Args:
+            lift_windows -- np_array (Np array of the lift windows)
+            rows -- int (# of rows within a trace (ex: x, y, z, nanos => 4 rows))
+            window_size -- int (# of samples within one window when creating the windows)
+            stride -- int (# of samples to jump before creating next window)
+            variance_threshold -- int (Variance threshold that determines
+                                  wether the current window is part of the lift)
+
+
+        Return:
+            start_time -- int (start time in nanos)
+            end -- int (end time in nanos)
+        """
+    start, end = find_initial_lift_times(lift_windows, rows)
+    precise_start = find_precise_start_time(
+        lift_windows, rows, window_size, stride, variance_threshold)
+    if (precise_start != None):
+        start_time = precise_start
+    else:
+        start_time = start
+    return start_time, end
+
+
+def centered_window(np_sample, center_time, window_size):
+    """
+    Description: Creates a window centered around the center_time of the lift
+
+    Args:
+        np_sample -- np array (The full trace as an np array)
+        center_time -- int (Time in nanoseconds that the window should be centered around)
+        window_size -- int (Size of window)
+
+
+    Return:
+        window -- np_array (Window centered around the given time)
+    """
+    nanos = np_sample[3]
+    center_index = np.where(nanos == center_time)[0][0]
+    add_front = int(window_size/2)
+    add_back = int(window_size/2)
+    if window_size%2 !=0:
+        add_back += 1
+    back_index = center_index + add_back
+    front_index = center_index - add_front
+    last = len(np_sample[0])-1
+    if (front_index < 0):
+        add_back += abs(front_index)
+        front_index = 0
+        back_index = center_index + add_back
+    if (back_index > last):
```

Comment:
Do you know how often this shifting occurs? I wonder how many of the windows are actually aligned together for the experiments? I'm assuming it's most of them and this is rare.

Reply:
It depends just on the individual cases for the experiments I ran it did not happen that often until the window size got pretty big.